### PR TITLE
[feat] Restake module

### PR DESCRIPTION
**Pull request by RogerKSI**
## Implementation details
- Implement restake module (lock/unlock power, stake/unstake liquid token to get more power)
- Add unit test

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [x] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: x/restake/keeper/keeper_test.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+		SetModuleAccount(gomock.Any(), moduleAccount).
+		Return().
+		AnyTimes()
+	suite.accountKeeper = accountKeeper
+
+	bankKeeper := restaketestutil.NewMockBankKeeper(ctrl)
+	bankKeeper.EXPECT().
+		GetAllBalances(gomock.Any(), moduleAccount.GetAddress()).
+		Return(nil).
+		Times(1)
```
Comment (colmazia) [inline x/restake/keeper/keeper_test.go:140]: Why just one time?

File: x/restake/keeper/keeper_vault.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+			key,
+			true,
+		)
+
+		k.SetVault(ctx, vault)
+
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				types.EventTypeCreateVault,
+				sdk.NewAttribute(types.AttributeKeyKey, key),
```
Comment (colmazia) [inline x/restake/keeper/keeper_vault.go:25]: AttributeKeyKey feels weird

File: x/restake/keeper/keeper_vault_test.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	ctx := suite.ctx
+	suite.setupState()
+
+	// error case -  no vault
+	err := suite.restakeKeeper.DeactivateVault(ctx, InvalidVaultKey)
+	suite.Require().Error(err)
+
+	// error case - vault is deactivated
+	err = suite.restakeKeeper.DeactivateVault(ctx, InactiveVaultKey)
+	suite.Require().Error(err)
```
Comment (colmazia) [inline x/restake/keeper/keeper_vault_test.go:52]: should we test the error message?

File: x/restake/keeper/msg_server_test.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+					sdk.NewCoin("uband", sdkmath.NewInt(2000)),
+				),
+			},
+			expErr:    true,
+			expErrMsg: "stake not enough",
+			preCheck:  func() {},
+			postCheck: func() {},
+		},
+		{
+			name: "success",
```
Comment (colmazia) [inline x/restake/keeper/msg_server_test.go:155]: should we add case of not enough total power after unstake

File: x/restake/keeper/genesis.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+		k.SetLock(ctx, lock)
+	}
+
+	var totalStakes sdk.Coins
+	for _, stake := range data.Stakes {
+		k.SetStake(ctx, stake)
+		totalStakes = totalStakes.Add(stake.Coins...)
+	}
+
+	// check if total stakes equals balances, if it doesn't panic because there were export/import errors
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply x/restake/keeper/genesis.go:42]: fixed.

File: x/restake/keeper/grpc_query.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+	"cosmossdk.io/store/prefix"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/query"
+
+	"github.com/bandprotocol/chain/v3/x/restake/types"
+)
+
+// Querier is used as Keeper will have duplicate methods if used directly, and gRPC names take precedence over keeper.
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply x/restake/keeper/grpc_query.go:17]: fixed

File: x/restake/keeper/keeper_test.go (hunk 1 above)
Comment (RogerKSI) [reply x/restake/keeper/keeper_test.go:140]: this expectation is for initgenesis on setup only. after that, the balance might be changed.

File: x/restake/keeper/msg_server_test.go (hunk 4 above)
Comment (RogerKSI) [reply x/restake/keeper/msg_server_test.go:155]: added

File: x/restake/keeper/keeper_vault_test.go (hunk 3 above)
Comment (RogerKSI) [reply x/restake/keeper/keeper_vault_test.go:52]: fixed

**Review by colmazia — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**